msa/js-executor/queue/kafkaTemplate.js

```diff
@@ -33,21 +35,75 @@ let producer;
 
 const configEntries = [];
 
+let batchMessages = [];
+let batchResolvers = [];
+let sendLoopInstance;
+
 function KafkaProducer() {
-    this.send = async (responseTopic, scriptId, rawResponse, headers) => {
-        return producer.send(
-            {
-                topic: responseTopic,
-                acks: acks,
-                compression: compressionType,
-                messages: [
-                    {
-                        key: scriptId,
-                        value: rawResponse,
-                        headers: headers.data
-                    }
-                ]
-            });
+    this.send = (responseTopic, scriptId, rawResponse, headers) => {
+        logger.debug('Pending queue response, scriptId: [%s]', scriptId);
+        const message = {
+            topic: responseTopic,
+            messages: [{
+                key: scriptId,
+                value: rawResponse,
+                headers: headers.data
+            }]
+        };
+
+        return pushMessageToSendLater(message);
+    }
+}
+
+function pushMessageToSendLater(message) {
+    let resolver;
+    const promise = new Promise((resolve, reject) => {
+        resolver = resolve;
+    });
+    batchMessages.push(message);
+    batchResolvers.push(resolver);
+    if (batchMessages.length >= maxBatchSize) {
+        sendMessagesAsBatch();
+        sendLoopWithLinger(); //reset loop function and reschedule new linger
+    }
+    return promise;
```

Comment:
Does the previous implementation also return promise? Does the consumer "wait" for the messages to be published before consuming the incoming message pack? I am not saying this need to be fix immediately, but let's add a "TODO".

Reply:
The previous implementation was async. The actual implementation is async too. We don't wait on any locks.
The only sync job is to prepare a batch array (if maxBatchSize) and send async as a batch.
Another event is a timer (controlled by linger parameter), async too.